byosdk.md

```diff
@@ -0,0 +1,667 @@
+      query_timezone: q.query_timezone || undefined,
+      row_total: q.row_total || undefined,
+      sorts: q.sorts || [],
+      subtotals: q.subtotals || undefined,
+      total: typeof q.total !== 'undefined' ? q.total : false,
+      view: q.view!,
+      vis_config: q.vis_config || undefined,
+      visible_ui_sections: q.visible_ui_sections || undefined,
+    },
+    result_format: 'csv',
+  }
+  const csvFile = './query.csv'
+  const writer = fs.createWriteStream(csvFile)
+  await sdk.stream.run_inline_query(async (readable: Readable) => {
+    return new Promise<any>((resolve, reject) => {
+      readable
+        .pipe(writer)
+        .on('error', reject)
+        .on('finish', resolve)
+    })
+  }, request)
+```
+
+#### Request parameter encoding
+
+This section discusses how parameters should be processed by the run-time **before** submitting the HTTP request to the endpoint.
+
+**Path parameters**
+- include all path parameters
+- url-encode the values for a path containing variables e.g., `/users/{user_id}`
+
+**Query parameters**
+- url-encode all values
+- date values should be formatted in UTC time format like "2020-03-15T13:16:34.692-07:00"
+- skip `null` or `undefined` values
+
+**Body object**
+- include all required properties (this is typically enforced by the declared method's interface)
+- skip properties that are optional and `null` or `undefined`
+
+### AuthSession implementations
+
+When the codegen project began, there was one method of authenticating for Looker API requests, which was providing the
+API credentials `client_id` and `client_secret` to the API. `AuthSession` is commonly the name of this class.
+Due to the complexities of Node vs. Browser implementations and authentication methods, the reference
+implementation for the Typescript SDK is called [`NodeSession`](packages/sdk/src/rtl/nodeSession.ts) but for this
+document we'll pretend it's called `AuthSession`.
+
+Recent (Looker 7.6 and above) API implementations additionally offer an OAuth authentication flow for API calls. Both 
+API3 credentials and OAuth2 authentication methods should be supported by `AuthSession` implementations. There will be 
+additional authentication methods in future Looker releases, but they should follow very similar patterns to these methods.
+
+What all `AuthSession` implementations have in common is:
+
+- the use of an authorization token (typically added into a request's properties by the `authentication` callback) to
+  authenticate the API user's request
+- the session tracks authorization token expiration. If it is expired, the session automatically re-authenticates the
+  API user by reading the necessary credential information from `readConfig()` then logs the user back in, discards
+  those retrieved credentials, and saves the new authorization token for subsequent API requests.
+  
+**NOTE**: The token expiration tracking should be set for 10 - 15 seconds before the indicated token expiration period
```

Comment:
Yeah, I responded to that issue and also invited Danny to chime in. I think token expiration is 60 minutes. 10% of that would be 6 minutes. I think that's way more fudge factor than we need.

Reply:
Totally agreed on the blind spots we may have about what we should document in that guide. I was even forgetting things I thought should be included **while** writing the document, so I expect this will get frequent updates when it's first used as a guide for writing a new SDK. Thanks for looking it over.